plugins/gnupgplugin/gpgmebase.cpp

```diff
@@ -0,0 +1,761 @@
+    return Result(Result::OperationNotSupportedError,
+                  QStringLiteral("The GnuPG plugin doesn't support generation of random data."));
+}
+
+Result Daemon::Plugins::GnuPGPlugin::generateInitializationVector(CryptoManager::Algorithm algorithm,
+                                                                  CryptoManager::BlockMode blockMode,
+                                                                  int keySize,
+                                                                  const QVariantMap &customParameters,
+                                                                  QByteArray *generatedIV)
+{
+    Q_UNUSED(algorithm);
+    Q_UNUSED(blockMode);
+    Q_UNUSED(keySize);
+    Q_UNUSED(customParameters);
+    Q_UNUSED(generatedIV);
+    return Result(Result::OperationNotSupportedError,
+                  QStringLiteral("The GnuPG plugin doesn't support generation of initialisation vector."));
+}
+
+Result Daemon::Plugins::GnuPGPlugin::generateKey(const Key &keyTemplate,
+                                                 const KeyPairGenerationParameters &kpgParams,
+                                                 const KeyDerivationParameters &skdfParams,
+                                                 const QVariantMap &customParameters,
+                                                 Key *key,
+                                                 const QString &home)
+{
+    Q_UNUSED(skdfParams);
+    Q_UNUSED(customParameters);
+
+    if (!kpgParams.isValid()) {
+        return Result(Result::CryptoPluginKeyGenerationError,
+                      QStringLiteral("invalid key generation parameters."));
+    }
+    if (kpgParams.keyPairType() != KeyPairGenerationParameters::KeyPairDsa
+        && kpgParams.keyPairType() != KeyPairGenerationParameters::KeyPairRsa
+        && kpgParams.keyPairType() != KeyPairGenerationParameters::KeyPairCustom) {
+        return Result(Result::CryptoPluginKeyGenerationError,
+                      QStringLiteral("unsupported key pair algorithm."));
+    }
+    QVariantMap::ConstIterator name = kpgParams.customParameters().constFind("name");
+    if (m_protocol == GPGME_PROTOCOL_OpenPGP
+        && name == kpgParams.customParameters().constEnd()) {
+        return Result(Result::CryptoPluginKeyGenerationError,
+                      QStringLiteral("missing name custom parameter."));
+    }
+    QVariantMap::ConstIterator email = kpgParams.customParameters().constFind("email");
+    if (email == kpgParams.customParameters().constEnd()) {
+        return Result(Result::CryptoPluginKeyGenerationError,
+                      QStringLiteral("missing email custom parameter."));
+    }
+
+    GPGmeContext ctx(m_protocol, home);
+    if (!ctx)
+        return Result(Result::CryptoPluginKeyGenerationError, ctx.error());
+
+    QString gnupgKeyParms = "<GnupgKeyParms format=\"internal\">\n";
+    if (kpgParams.keyPairType() == KeyPairGenerationParameters::KeyPairDsa) {
```

Comment:
in this case, you could do: DsaKeyPairGenerationParameters dkpgp(kpgParams); and then read the key-length from the modulusLength() property of the dkpgp instance.

Reply:
In fact, for DSA, GnuPG handle only length of 1024. I've added a warning in case the user would have chosen another value.